### PR TITLE
backupccl: add incremental_storage option to RESTORE

### PR DESCRIPTION
**Pull request by msbutler**
Previously, the user could not restore incremental backups created with BACKUP
... WITH incremental_storage. This change adds the 'incremental_storage' option
to RESTORE, allowing users to restore a [locality aware] incremental backup
chain stored in 'incremental_storage'.

One UI subtlety: when creating a locality aware incremental backup with
'incremental_storage', users don't need to specify the localities of the full
backup, i.e.:

BACKUP INTO LATEST IN
'path_to_full_backup',
WITH incremental_storage=('path_to_inc_backup?COCKROACH_LOCALITY=default',
'path_to_inc_backup?COCKROACH_LOCALITY=region%3Dus-west');

However, users must do so in RESTORE:

RESTORE foo FROM LATEST IN ('path_to_full_backup?COCKROACH_LOCALITY=default',
'path_to_full_backup?COCKROACH_LOCALITY=region%3Dus-west') WITH
incremental_storage = ('path_to_inc_backup?COCKROACH_LOCALITY=default',
'path_to_inc_backup?COCKROACH_LOCALITY=region%3Dus-west')

Informs: #72236

Release note (sql change): users can now RESTORE [locality aware] incremental
backups created with the 'incremental_storage' parameter.

### CODE REVIEW

File: pkg/ccl/backupccl/backup_planning.go [hunk 1]
```diff
@@ -333,6 +333,7 @@ func resolveOptionsForBackupJobDescription(
 	}
 
 	var err error
+	// MB note: should we be using cloud.RedactKMSURI(uri) here instead?
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:336]: nit: if you want to stick your name/initials on it, may as well use the standard `TODO(<name>)` syntax

File: pkg/sql/sem/tree/backup.go [hunk 2]
```diff
@@ -139,10 +140,22 @@ var _ NodeFormatter = &RestoreOptions{}
...
-	Options            RestoreOptions
-	Subdir             Expr
+
+	// From contains the URIs for the backup(s) we seek to restore.
+	//   - len(From)>1 implies the user explicitly passed incremental backup paths,
+	//     which is only allowed using the old syntax, `RESTORE <targets> FROM <destination>.
+	//     In this case, From[0] contains the URI(s) for the full backup.
+	//   - len(From)==1 implies we'll have to look for incremental backups in planning
+	//   - len(From[0]) > 1 implies the backups are locality aware
+	//   - From[i][0] must be the default locality.
```
Comment (dt) [inline pkg/sql/sem/tree/backup.go:150]: Is this true elsewhere? i.e. during BACKUP can the =default locality be in any position?

**Review by dt — APPROVED** [review]

File: pkg/sql/sem/tree/backup.go (hunk 2 above)
Comment (msbutler) [reply pkg/sql/sem/tree/backup.go:150]: it's not true for BACKUP, as the default locality there is resolved using `getURIsByLocalityKV` in backup_planning.go, whereas in RESTORE, `resolveBackupManifests` assumes the first uri is default.... I'll create an issue related to refactoring destination resolution code in B/R planning.

File: pkg/ccl/backupccl/backup_planning.go (hunk 1 above)
Comment (msbutler) [reply pkg/ccl/backupccl/backup_planning.go:336]: Done

**Comment by msbutler** [comment]
TYFR!!

**Comment by msbutler** [comment]
bors r=dt

**Pull request merged**